File: src/go/finished/longest_subarray_with_maximum_bitwise_and.go
package leetcode

// Time: O(N)
// Space: O(1)
func longestSubarray(nums []int) int {
	N := len(nums)

	best := 0
	bestVal := 0
	for i := 0; i < N; {
		tmp := nums[i]
		count := 1
		for j := i + 1; j < N && nums[j] == tmp; j++ {
			count++
		}
		if tmp == bestVal {
			best = max(best, count)
		} else if tmp > bestVal {
			best = count
			bestVal = tmp
		}
		i += count
	}

	return best
}

// func max(a, b int) int {
// 	if a > b {
// 		return a
// 	}
// 	return b
// }

File: src/python/finished/longest_subarray_with_maximum_bitwise_and.py
from typing import List

# https://leetcode.com/problems/longest-subarray-with-maximum-bitwise-and/description


# O(n) time
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        N = len(nums)
        best = 0
        best_num = -1
        i = 0
        while i < N:
            n = nums[i]
            llen = 0
            while i < N and n == nums[i]:
                llen += 1
                i += 1
            if n == best_num:
                best = max(best, llen)
            elif n > best_num:
                best = llen
                best_num = n
        return best
